### PR TITLE
Add prevent-proxy-connections option

### PR DESCRIPTION
**Pull request by mikroskeem**
Addresses #296 

This PR is basically ready, but I couldn't come up with decent description for the option as there isn't *that* much information available for this in my opinion.

Perhaps option name could be different too?

### CODE REVIEW

**Comment by Cybermaxke** [comment]
Maybe a bit more specific? `prevent-client-proxy-connections`?

**Comment by mikroskeem** [comment]
That's better and reduces confusion. Now only thing to figure out is proper description for the option

**Comment by hugmanrique** [comment]
The Minecraft Wiki [says the following](https://minecraft.gamepedia.com/Server.properties):

> If the ISP/AS sent from the server is different from the one from Mojang's authentication server, the player is kicked
true - Enabled. Server prevents users from using vpns or proxies.
false - Disabled. The server doesn't prevent users from using vpns or proxies.

**Comment by mikroskeem** [comment]
Ready for proper reviewing now.

File: api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java [hunk 1]
```diff
@@ -62,6 +62,13 @@
    */
   boolean isOnlineMode();
 
+  /**
+   * TODO
```
Comment (hugmanrique) [inline api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java:66]: Mind adding the same description here?

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java [hunk 2]
```diff
@@ -54,6 +54,14 @@
   @ConfigKey("online-mode")
   private boolean onlineMode = true;
 
+  @Comment({
+      "If client's ISP/AS sent from this proxy is different from the one from Mojang's",
+      "authentication server, the player is kicked. Effectively disallows players from",
+      "using proxies or VPNs when enabled."
+  })
+  @ConfigKey("prevent-client-proxy-connections")
+  private boolean preventClientProxyConnections = false;
```
Comment (hugmanrique) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java:63]: Previous code defaulted to passing the IP param, should this be `true` instead?

Comment (mikroskeem) [reply proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java:63]: Vanilla has this off by default, so I went with false.

File: api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java (hunk 1 above)
Comment (mikroskeem) [reply api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java:66]: Oops

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java (hunk 2 above)
Comment (astei) [reply proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java:63]: I'm inclined to say we should leave this off, as `prevent-proxy-connections` is a rather weak form of protection that can be easily bypassed.

File: api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java [hunk 3]
```diff
@@ -62,6 +62,15 @@
    */
   boolean isOnlineMode();
 
+  /**
+   * If client's ISP/AS sent from this proxy is different from the one from Mojang's"
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java:66]: Errant `"`

File: api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java [hunk 4]
```diff
@@ -62,6 +62,15 @@
...
+   * authentication server, the player is kicked. Effectively disallows players from
+   * using proxies or VPNs when enabled.
+   *
+   * @return prevent client proxy connections
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java:70]: Make this a little bit more explanatory

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java [hunk 5]
```diff
@@ -96,8 +96,11 @@ public boolean handle(EncryptionResponse packet) {
 
       String playerIp = ((InetSocketAddress) mcConnection.getRemoteAddress()).getHostString();
       String url = String.format(MOJANG_HASJOINED_URL,
-          urlFormParameterEscaper().escape(login.getUsername()), serverId,
-          urlFormParameterEscaper().escape(playerIp));
+          urlFormParameterEscaper().escape(login.getUsername()), serverId);
+
+      if (server.getConfiguration().shouldPreventClientProxyConnections()) {
+        url += urlFormParameterEscaper().escape("&ip=" + playerIp);
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java:102]: Not sure if we need to escape the `&ip=` here. We probably should escape the IP but that's it.

File: api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java (hunk 3 above)
Comment (mikroskeem) [reply api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java:66]: Yeah copied from the configuration class, oops.

File: api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java [hunk 6]
```diff
@@ -62,6 +62,15 @@
    */
   boolean isOnlineMode();
 
+  /**
+   * If client's ISP/AS sent from this proxy is different from the one from Mojang's"
+   * authentication server, the player is kicked. Effectively disallows players from
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/proxy/config/ProxyConfig.java:67]: "Effectively" is too strong. I'd do "This disallows some VPNs and proxy connections but is a weak form of protection."

**Pull request merged**